### PR TITLE
Docstring for `Int` and its basic operations

### PR DESCRIPTION
**Pull request by AdrienChampion**
Mentored by @digama0 

- add a docstring for `Int`, including a discussion of the special internal representation;
- document basic operations with examples, in particular non-standard cases, *e.g.* `n / 0 = 0`.

Unfortunately, there is no discussion of (asymptotic) complexity. My understanding is that the issue is quite complex, as operations over "bignums" call GMP (or even something else?) and it's not always clear exactly how the operation will be performed at runtime.

I [attempted to get feedback/directions on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Int.20operations.20complexity) without any success.

### CODE REVIEW

File: src/Init/Data/Int/Basic.lean [hunk 1]
```diff
@@ -28,32 +55,69 @@ instance : OfNat Int n where
...
+  Implemented by efficient native code. -/
 @[extern "lean_int_neg"]
 protected def neg (n : @& Int) : Int :=
   match n with
   | ofNat n   => negOfNat n
   | negSucc n => succ n
 
+/--
+  The `Neg Int` default instance must have priority higher than `low`
+  since the default instance `OfNat Nat n` has `low` priority.
```
Comment (digama0) [inline src/Init/Data/Int/Basic.lean:75]: this should be a regular comment, not a doc comment. Doc comments on instances are mostly useless to begin with

File: src/Init/Data/Int/Basic.lean [hunk 2]
```diff
@@ -9,12 +9,39 @@ prelude
 import Init.Coe
 import Init.Data.Nat.Div
 import Init.Data.List.Basic
+set_option linter.missingDocs true -- keep it documented
 open Nat
 
-/-! # the Type, coercions, and notation -/
+/-! # `Int`eger Type, Coercions, and Notation
```
Comment (gebner) [inline src/Init/Data/Int/Basic.lean:15]:
```suggestion
/-! # Integer Type, Coercions, and Notation
```

File: src/Init/Data/Int/Basic.lean [hunk 3]
```diff
@@ -9,12 +9,39 @@ prelude
...
+* basic arithmetic operations add/sub/mul/div/mod/pow,
+* a few `Nat`-related operations such as `negOfNat` and `subNatNat`,
+* relations `<`/`≤`/`≥`/`>`, the `NonNeg` property and `min`/`max`,
+* decidability of equality, relations and `NonNeg`.
+-/
+
+/--
+The type of integers. It is defined as an inductive type based on the
+natural number type `Nat` featuring two constructors: "a natural
+number is an integer", and "the negation of a `Nat.succ`essor of a
```
Comment (gebner) [inline src/Init/Data/Int/Basic.lean:29]:
```suggestion
number is an integer", and "the negation of a successor of a
```

**Review by gebner — COMMENTED** [review]
Please proof-read the text.

EDIT: or are these intentional?  We certainly don't write `Int`eger anywhere else.

**Comment by digama0** [comment]
I think it's intentional, but I agree it is a bit too "cute" to keep.

**Comment by AdrienChampion** [comment]
Yes, it was intentional. I'm used to this kind of "cute" formatting, but if it's not used anywhere else we might as well remove it.

File: src/Init/Data/Int/Basic.lean [hunk 4]
```diff
@@ -113,51 +187,162 @@ protected def decEq (a b : @& Int) : Decidable (a = b) :=
...
+  ```
+
+  Implemented by efficient native code. -/
 @[extern "lean_int_dec_nonneg"]
 private def decNonneg (m : @& Int) : Decidable (NonNeg m) :=
   match m with
   | ofNat m   => isTrue <| NonNeg.mk m
   | negSucc _ => isFalse <| fun h => nomatch h
 
+/-- Decides wether `a ≤ b`.
```
Comment (Kha) [inline src/Init/Data/Int/Basic.lean:205]:
```suggestion
/-- Decides whether `a ≤ b`.
```

**Review by Kha — APPROVED** [review]

File: src/Init/Data/Int/Basic.lean [hunk 5]
```diff
@@ -62,42 +126,52 @@ protected def mul (m n : @& Int) : Int :=
...
 
 instance : Sub Int where
   sub := Int.sub
 
+/-- A proof that an `Int` is non-negative. -/
 inductive NonNeg : Int → Prop where
+  /-- Sole constructor, proving that `ofNat n` is positive. -/
   | mk (n : Nat) : NonNeg (ofNat n)
 
+/-- Proof that `a ≤ b` encoded as `b - a ≥ 0`. -/
```
[the remark this answers is not in the document]
Comment (AdrienChampion) [reply src/Init/Data/Int/Basic.lean]: Fixed, also fixed the same mistake in `a < b`

File: src/Init/Data/Int/Basic.lean (hunk 4 above)
Comment (AdrienChampion) [reply src/Init/Data/Int/Basic.lean]: Fixed, also fixed for `a < b`

**Comment by bustercopley** [comment]
@AdrienChampion, the comment above `Int.div` has:
```
  #eval (12 : Int) / (7 : Int) -- 1
  #eval (12 : Int) / (-7 : Int) -- -1
  #eval (-12 : Int) / (7 : Int) -- -2
  #eval (-12 : Int) / (-7 : Int) -- 2
```
but the correct results (of both the builtin and the `def` given there) are as follows:
```lean4
import Init.Data.Int

open Int

def Int.tdiv : (@& Int) → (@& Int) → Int
  | ofNat m,   ofNat n   => ofNat (m / n)
  | ofNat m,   negSucc n => -ofNat (m / Nat.succ n)
  | negSucc m, ofNat n   => -ofNat (Nat.succ m / n)
  | negSucc m, negSucc n => ofNat (Nat.succ m / Nat.succ n)

#eval (12 : Int) / (7 : Int)      --  1
#eval (12 : Int) / (-7 : Int)     -- -1
#eval (-12 : Int) / (7 : Int)     -- -1
#eval (-12 : Int) / (-7 : Int)    --  1

#eval (12 : Int).div (7 : Int)    --  1
#eval (12 : Int).div (-7 : Int)   -- -1
#eval (-12 : Int).div (7 : Int)   -- -1
#eval (-12 : Int).div (-7 : Int)  --  1

#eval (12 : Int).tdiv (7 : Int)   --  1
#eval (12 : Int).tdiv (-7 : Int)  -- -1
#eval (-12 : Int).tdiv (7 : Int)  -- -1
#eval (-12 : Int).tdiv (-7 : Int) --  1
```

**Pull request merged**